### PR TITLE
Add Dirichlet Weight Initialization

### PR DESCRIPTION
**Pull request by jonpsy**
Continuing #293 

- [x] Added dirichilet_init.hpp
- [x] Test and verify locally.
- [x] ```using DirichletMOEAD = MOEAD<Dirichlet, Tchebycheff>```
- [x] Document in dirichlet_init.hpp
- [x] Add documentation in optimizers.md
- [x] History.md

### CODE REVIEW

File: include/ensmallen_bits/moead/weight_init_policies/dirichlet_init.hpp [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+#define ENSMALLEN_MOEAD_DIRICHLET_HPP
+
+namespace ens {
+
+/**
+ * The Dirichlet method for initializing weights. Sampling a dirichlet distribution
+ * with parameters set to one returns points lying on unit simplex with uniform
+ * distribution.
+ *
+ */
```
Comment (jonpsy) [inline include/ensmallen_bits/moead/weight_init_policies/dirichlet_init.hpp:22]: @zoq  Not sure what should be the source ?

Comment (zoq) [reply include/ensmallen_bits/moead/weight_init_policies/dirichlet_init.hpp:22]: I don't think there is one we should reference here.

**Review by zoq — COMMENTED** [review]
Can you add a test for the policy as well?

**Comment by jonpsy** [comment]
You have any specifics in mind?

**Comment by zoq** [comment]
We can just use the exciting one but change the policy.

**Comment by jonpsy** [comment]
I could test it against a difficult ZDT problem. See how it fares? The paper has tried testing against some ZDT problems. We can see how much the theoretical results match.

**Comment by zoq** [comment]
Right, let's test it one e.g. one ZDT problem.

File: include/ensmallen_bits/moead/weight_init_policies/dirichlet_init.hpp [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+   */
+  template<typename MatType>
+  MatType Generate(const size_t numObjectives,
+                   const size_t numPoints,
+                   const double epsilon)
+  {
+    MatType weights = arma::randg<MatType>(numObjectives, numPoints,
+        arma::distr_param(1.0, 1.0)) + epsilon;
+    // Normalize each column.
+    return arma::normalise(weights, 1, 0);
```
Comment (coatless) [inline include/ensmallen_bits/moead/weight_init_policies/dirichlet_init.hpp:50]: So, thinking about this... We have a gamma distribution given as: 

```
Y_i ~ gamma(alpha_i, 1)
```

From there, we're creating the normalization of:

```
X = sum_{i = 1}^{n} Y_i
V = (Y_1/X, Y_2/X, ... Y_n/X) ~ Dirchlet(alpha_1, ... alpha_n)
```
https://en.wikipedia.org/wiki/Dirichlet_distribution#Gamma_distribution

Looking at: 

http://arma.sourceforge.net/docs.html#normalise

`normalise` is giving is a unit _p_-norm. With `p = 1`, we're getting: 

```
X = sum abs(Y_i)
```

The `abs()` goes against the summation component, but gamma RVs will always be > 0.

So, long story short... We're doing one extra operation, but the end outcome will be the same.

**Review by coatless — APPROVED** [review]

File: include/ensmallen_bits/moead/weight_init_policies/dirichlet_init.hpp [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+   * @param numObjectives The dimensionality of objective space.
+   * @param numPoints The number of reference directions requested.
+   * @param epsilon Handle numerical stability after weight initialization.
+   */
+  template<typename MatType>
+  MatType Generate(const size_t numObjectives,
+                   const size_t numPoints,
+                   const double epsilon)
+  {
+    MatType weights = arma::randg<MatType>(numObjectives, numPoints,
```
[the remark this answers is not in the document]
Comment (jonpsy) [reply include/ensmallen_bits/moead/weight_init_policies/dirichlet_init.hpp:47]: alpha = 1. See [this](https://cs.stackexchange.com/a/132116).

**Comment by jonpsy** [comment]
Planning to add
```using DirichletMOEAD = MOEAD<Dirichlet, Tchebycheff>```

File: include/ensmallen_bits/moead/weight_init_policies/dirichlet_init.hpp (hunk 2 above)
Comment (jonpsy) [reply include/ensmallen_bits/moead/weight_init_policies/dirichlet_init.hpp:50]: I think ```p=1``` gives something like

```std::sqrt( elem0^2 + elem1^2 + elem2^2 ..... elemn^2)``` and not the sum if I've gotten this correctly. I suppose you're saying, assuming X = sum abs(Y_i); abs part is unnecessary?

**Comment by jonpsy** [comment]
> Right, let's test it one e.g. one ZDT problem.

I tested it against ZDT3 problem. Ran it over 1000 times using random seed using this bash command

```bash
#!/bin/bash
for (( i = 0; i < 1000; i++))
do
 ./ensmallen_tests MOEADDIRICHLETZDT3Test --rng-seed time
done
```

File: include/ensmallen_bits/moead/weight_init_policies/dirichlet_init.hpp (hunk 2 above)
Comment (coatless) [reply include/ensmallen_bits/moead/weight_init_policies/dirichlet_init.hpp:49]: Nope, that would be `p=2`/ l2-norm. This is using an l1-norm by column.

`p=1` gives: 

```cpp
std::pow( abs(elem0)^1 + abs(elem1)^1 + abs(elem2)^1 ..... abs(elemn)^1, 1.0);
```
or:

```
(abs(elem0)^1 + abs(elem1^1 + abs(elem2^1  + ..... + abs(elemn^1))^1 
= abs(elem0)^1 + abs(elem1)^1 + abs(elem2)^1  + ..... +  abs(elemn)^1 
= abs(elem0) + abs(elem1) + abs(elem2) + ..... + abs(elemn)
```

 I think by-column is correct instead of by-row due to the weights being structured `numObjectives x numPoints` (norm across objectives under a single point)

File: include/ensmallen_bits/moead/weight_init_policies/dirichlet_init.hpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (coatless) [reply include/ensmallen_bits/moead/weight_init_policies/dirichlet_init.hpp:46]: Okay, so this weight policy is being designed with a specific application in mind so that `alpha` doesn't need to be flexible.

[the remark this answers is not in the document]
Comment (jonpsy) [reply include/ensmallen_bits/moead/weight_init_policies/dirichlet_init.hpp:46]: Yes, since we want uniform distribution. The alphas vectors are set to ones vector.

File: include/ensmallen_bits/moead/weight_init_policies/dirichlet_init.hpp (hunk 2 above)
Comment (jonpsy) [reply include/ensmallen_bits/moead/weight_init_policies/dirichlet_init.hpp:49]: Yes, I've done it in a column basis so the normalization of each point is ensured. Just revised L-norms, I was mistaken about the L1 norm formula, thanks for pointing it out.

I think the "abs" part shouldn't really be that of a concern, either way, we know each element is > 0. We have one line of code which do the thing for us, and I really don't think the additional "abs" would severely hinder the performance.

**Comment by jonpsy** [comment]
I suppose it's fair to say this PR is GTM?

**Review by zoq — APPROVED** [review]
Looks great to me.

**Comment by zoq** [comment]
Let's see if the build comes back green and merge it afterwards.

**Comment by zoq** [comment]
Thanks :+1:
